File: app/src/main/java/br/com/menuvem/lojista/presentation/products/ProdutosScreen.kt
package br.com.menuvem.lojista.presentation.products

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material.icons.outlined.Restaurant
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import br.com.menuvem.lojista.domain.model.ProdutoComCusto
import br.com.menuvem.lojista.presentation.components.EmptyState
import br.com.menuvem.lojista.presentation.components.formatarMoeda
import br.com.menuvem.lojista.ui.theme.*

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ProdutosScreen(
    onNavigateToProduto: (Long) -> Unit,
    onBack: () -> Unit,
    viewModel: ProdutosViewModel = hiltViewModel()
) {
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Produtos") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = "Voltar"
                        )
                    }
                }
            )
        },
        floatingActionButton = {
            ExtendedFloatingActionButton(
                onClick = viewModel::onShowForm,
                icon = { Icon(Icons.Default.Add, contentDescription = null) },
                text = { Text("Novo Produto") },
                containerColor = YellowSecondary,
                contentColor = MaterialTheme.colorScheme.surface
            )
        }
    ) { paddingValues ->
        when {
            uiState.isLoading -> Box(
                modifier = Modifier.fillMaxSize().padding(paddingValues),
                contentAlignment = Alignment.Center
            ) { CircularProgressIndicator(color = PurplePrimary) }

            uiState.produtos.isEmpty() -> Box(modifier = Modifier.padding(paddingValues)) {
                EmptyState(
                    icon = Icons.Outlined.Restaurant,
                    title = "Nenhum produto cadastrado",
                    subtitle = "Cadastre um produto e monte a ficha técnica para calcular o custo e o preço ideal"
                )
            }

            else -> LazyColumn(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(paddingValues),
                contentPadding = PaddingValues(
                    start = 16.dp, top = 8.dp, end = 16.dp, bottom = 88.dp
                ),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                items(uiState.produtos, key = { it.produto.id }) { item ->
                    ProdutoCard(
                        item = item,
                        onClick = { onNavigateToProduto(item.produto.id) }
                    )
                }
            }
        }
    }

    if (uiState.showFormDialog) {
        ProdutoFormDialog(
            onDismiss = viewModel::onHideForm,
            onConfirm = { nome, margem, preco ->
                viewModel.createProduto(nome, margem, preco, onCreated = onNavigateToProduto)
            }
        )
    }
}

@Composable
private fun ProdutoCard(item: ProdutoComCusto, onClick: () -> Unit) {
    ElevatedCard(
        onClick = onClick,
        modifier = Modifier.fillMaxWidth(),
        elevation = CardDefaults.elevatedCardElevation(defaultElevation = 1.dp)
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = item.produto.nome,
                        style = MaterialTheme.typography.titleSmall,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis
                    )
                    Text(
                        text = if (item.quantidadeItens == 0) "Ficha técnica vazia"
                        else "${item.quantidadeItens} insumo(s) na ficha",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
                if (item.margemAbaixoDaMeta) {
                    Icon(
                        imageVector = Icons.Default.Warning,
                        contentDescription = "Margem abaixo da meta",
                        tint = WarningOrange,
                        modifier = Modifier.size(20.dp)
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                }
                Column(horizontalAlignment = Alignment.End) {
                    Text(
                        text = "Custo R$ ${formatarMoeda(item.custoTotal)}",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                    Text(
                        text = "R$ ${formatarMoeda(item.precoSugerido)}",
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold,
                        color = PurplePrimary
                    )
                    Text(
                        text = "preço sugerido",
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }

            Spacer(modifier = Modifier.height(10.dp))
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                MargemChip(
                    label = "Meta ${formatarPercentual(item.produto.margemAlvoPercentual)}",
                    container = PurpleContainer,
                    content = PurpleOnContainer
                )
                val margemReal = item.margemRealPercentual
                if (margemReal != null) {
                    MargemChip(
                        label = "Atual ${formatarPercentual(margemReal)}",
                        container = if (item.margemAbaixoDaMeta) WarningContainer else SuccessContainer,
                        content = if (item.margemAbaixoDaMeta) WarningOrange else SuccessGreen
                    )
                } else {
                    MargemChip(
                        label = "Sem preço de venda",
                        container = MaterialTheme.colorScheme.surfaceVariant,
                        content = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }
        }
    }
}

@Composable
private fun MargemChip(label: String, container: androidx.compose.ui.graphics.Color, content: androidx.compose.ui.graphics.Color) {
    Surface(shape = RoundedCornerShape(50), color = container) {
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = content,
            modifier = Modifier.padding(horizontal = 10.dp, vertical = 4.dp)
        )
    }
}

private fun formatarPercentual(valor: Double): String =
    String.format("%.1f%%", valor)

File: lib/presentation/products/produtos_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';

import '../../domain/model/produto_com_custo.dart';
import '../../domain/usecase/cost_engine.dart';
import '../../domain/usecase/produto_usecases.dart' as usecase;
import '../../providers/repository_providers.dart';
import '../../theme/app_theme.dart';
import '../components/back_or_home_button.dart';
import '../components/emoji_picker_field.dart';
import '../components/empty_state.dart';
import '../components/formatters.dart';
import '../components/responsive.dart';
import '../components/tutorial/tutorial_button.dart';
import '../components/tutorial/tutorial_keys.dart';
import '../components/tutorial/tutorial_passos.dart';
import 'produto_form_dialog.dart';

final produtosComCustoProvider = StreamProvider<List<ProdutoComCusto>>((ref) {
  return getProdutosComCusto(
    produtoRepository: ref.watch(produtoRepositoryProvider),
    insumoRepository: ref.watch(insumoRepositoryProvider),
    componenteRepository: ref.watch(componenteRepositoryProvider),
  );
});

class ProdutosScreen extends ConsumerWidget {
  const ProdutosScreen({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final async = ref.watch(produtosComCustoProvider);

    return Scaffold(
      appBar: AppBar(
        leading: const BackOrHomeButton(),
        title: const Text('Produtos'),
        actions: const [TutorialButton(tela: TutorialTela.produtos)],
      ),
      floatingActionButton: FloatingActionButton.extended(
        key: TutorialKeys.produtosNovo,
        onPressed: () => showResponsiveFormSheet<void>(
          context,
          builder: (_) => ProdutoFormDialog(
            onConfirm: (nome, emoji, margem, preco) async {
              // Todo produto nasce com uma porção "Única" — quem vende
              // tamanho único nunca precisa saber que porções existem.
              final (produtoId, _) = await usecase.criarProdutoComPorcaoInicial(
                ref.read(produtoRepositoryProvider),
                nome: nome,
                emoji: emoji,
                margemAlvoPercentual: margem,
                precoVendaAtual: preco,
              );
              if (context.mounted) context.push('/produtos/$produtoId');
            },
          ),
        ),
        backgroundColor: yellowSecondary,
        foregroundColor: Colors.white,
        icon: const Icon(Icons.add),
        label: const Text('Novo Produto'),
      ),
      body: MaxWidthCenter(
        child: async.when(
          loading: () => const Center(
            child: CircularProgressIndicator(color: purplePrimary),
          ),
          error: (e, _) => Center(child: Text('Erro: $e')),
          data: (produtos) {
            if (produtos.isEmpty) {
              return const EmptyState(
                icon: Icons.restaurant_outlined,
                title: 'Nenhum produto cadastrado',
                subtitle:
                    'Cadastre um produto e monte a ficha técnica para calcular o custo e o preço ideal',
              );
            }
            return ListView.separated(
              key: TutorialKeys.produtosLista,
              padding: const EdgeInsets.fromLTRB(16, 8, 16, 88),
              itemCount: produtos.length,
              separatorBuilder: (_, _) => const SizedBox(height: 8),
              itemBuilder: (context, i) => _ProdutoCard(
                item: produtos[i],
                onTap: () =>
                    context.push('/produtos/${produtos[i].produto.id}'),
              ),
            );
          },
        ),
      ),
    );
  }
}

class _ProdutoCard extends StatelessWidget {
  const _ProdutoCard({required this.item, required this.onTap});
  final ProdutoComCusto item;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final principal = item.porcaoPrincipal;
    final multiplas = item.temMultiplasPorcoes;
    final margemReal = principal?.margemRealPercentual;

    // Com vários tamanhos, custo e preço viram faixa (menor–maior).
    final custoLabel = multiplas
        ? 'Custo R\$ ${formatarMoeda(item.custoMinimo)}–${formatarMoeda(item.custoMaximo)}'
        : 'Custo R\$ ${formatarMoeda(principal?.custoTotal ?? 0)}';
    final precoLabel = multiplas
        ? 'R\$ ${formatarMoeda(item.precoSugeridoMinimo)}–${formatarMoeda(item.precoSugeridoMaximo)}'
        : 'R\$ ${formatarMoeda(principal?.precoSugerido ?? 0)}';

    return Card(
      child: InkWell(
        onTap: onTap,
        child: Padding(
          padding: const EdgeInsets.all(16),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Row(
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  EmojiAvatar(
                    emoji: item.produto.emoji,
                    fallback: Icons.restaurant_outlined,
                  ),
                  const SizedBox(width: 12),
                  Expanded(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          item.produto.nome,
                          maxLines: 1,
                          overflow: TextOverflow.ellipsis,
                          style: Theme.of(context).textTheme.titleSmall,
                        ),
                        Text(
                          multiplas
                              ? '${item.porcoes.length} tamanhos · ${item.quantidadeItens} insumo(s)'
                              : (item.quantidadeItens == 0
                                    ? 'Ficha técnica vazia'
                                    : '${item.quantidadeItens} insumo(s) na ficha'),
                          style: Theme.of(context).textTheme.bodySmall
                              ?.copyWith(
                                color: Theme.of(
                                  context,
                                ).colorScheme.onSurfaceVariant,
                              ),
                        ),
                      ],
                    ),
                  ),
                  if (item.algumaPorcaoAbaixoDaMeta)
                    const Padding(
                      padding: EdgeInsets.only(right: 8),
                      child: Icon(
                        Icons.warning,
                        color: warningOrange,
                        size: 20,
                      ),
                    ),
                  Column(
                    crossAxisAlignment: CrossAxisAlignment.end,
                    children: [
                      Text(
                        custoLabel,
                        style: Theme.of(context).textTheme.bodySmall?.copyWith(
                          color: Theme.of(context).colorScheme.onSurfaceVariant,
                        ),
                      ),
                      Text(
                        precoLabel,
                        style: Theme.of(context).textTheme.titleMedium
                            ?.copyWith(
                              fontWeight: FontWeight.bold,
                              color: purplePrimary,
                            ),
                      ),
                      Text(
                        'preço sugerido',
                        style: Theme.of(context).textTheme.labelSmall?.copyWith(
                          color: Theme.of(context).colorScheme.onSurfaceVariant,
                        ),
                      ),
                    ],
                  ),
                ],
              ),
              const SizedBox(height: 10),
              Wrap(
                spacing: 8,
                runSpacing: 4,
                children: multiplas
                    ? [
                        // Um chip por tamanho, destacando os que furaram a meta.
                        for (final p in item.porcoes)
                          _MargemChip(
                            label:
                                '${p.porcao.nome} · R\$ ${formatarMoeda(p.precoSugerido)}',
                            container: p.margemAbaixoDaMeta
                                ? warningContainer
                                : purpleContainer,
                            content: p.margemAbaixoDaMeta
                                ? warningOrange
                                : purpleOnContainer,
                          ),
                      ]
                    : [
                        _MargemChip(
                          label:
                              'Meta ${(principal?.porcao.margemAlvoPercentual ?? 0).toStringAsFixed(1)}%',
                          container: purpleContainer,
                          content: purpleOnContainer,
                        ),
                        if (margemReal != null)
                          _MargemChip(
                            label: 'Atual ${margemReal.toStringAsFixed(1)}%',
                            container: principal!.margemAbaixoDaMeta
                                ? warningContainer
                                : successContainer,
                            content: principal.margemAbaixoDaMeta
                                ? warningOrange
                                : successGreen,
                          )
                        else
                          _MargemChip(
                            label: 'Sem preço de venda',
                            container: Theme.of(
                              context,
                            ).colorScheme.surfaceContainerHighest,
                            content: Theme.of(
                              context,
                            ).colorScheme.onSurfaceVariant,
                          ),
                      ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class _MargemChip extends StatelessWidget {
  const _MargemChip({
    required this.label,
    required this.container,
    required this.content,
  });
  final String label;
  final Color container;
  final Color content;

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 4),
      decoration: BoxDecoration(
        color: container,
        borderRadius: BorderRadius.circular(50),
      ),
      child: Text(
        label,
        style: Theme.of(context).textTheme.labelSmall?.copyWith(color: content),
      ),
    );
  }
}
